backends/ebpf/ebpfDeparser.cpp

```diff
@@ -0,0 +1,376 @@
+                                          const IR::Expression* hdrExpr, unsigned int alignment,
+                                          EBPF::EBPFType* type) {
+    auto program = deparser->program;
+
+    auto et = dynamic_cast<EBPF::IHasWidth *>(type);
+    if (et == nullptr) {
+        ::error(ErrorType::ERR_UNSUPPORTED_ON_TARGET,
+                "Only headers with fixed widths supported %1%", hdrExpr);
+        return;
+    }
+    unsigned widthToEmit = et->widthInBits();
+    unsigned loadSize = 0;
+    cstring swap = "", msgStr;
+
+    if (widthToEmit <= 64) {
+        cstring tmp = Util::printf_format("(unsigned long long) %s.%s",
+                                          hdrExpr->toString(), field);
+        msgStr = Util::printf_format("Deparser: emitting field %s=0x%%llx (%u bits)",
+                                     field, widthToEmit);
+        builder->target->emitTraceMessage(builder, msgStr.c_str(), 1, tmp.c_str());
+    } else {
+        msgStr = Util::printf_format("Deparser: emitting field %s (%u bits)", field, widthToEmit);
+        builder->target->emitTraceMessage(builder, msgStr.c_str());
+    }
+
+    if (widthToEmit <= 8) {
+        loadSize = 8;
+    } else if (widthToEmit <= 16) {
+        swap = "bpf_htons";
+        loadSize = 16;
+    } else if (widthToEmit <= 32) {
+        swap = "htonl";
+        loadSize = 32;
+    } else if (widthToEmit <= 64) {
+        swap = "htonll";
+        loadSize = 64;
+    }
+    unsigned bytes = ROUNDUP(widthToEmit, 8);
+    unsigned shift = widthToEmit < 8 ?
+                     (loadSize - alignment - widthToEmit) : (loadSize - widthToEmit);
+
+    if (!swap.isNullOrEmpty()) {
+        builder->emitIndent();
+        visit(hdrExpr);
+        builder->appendFormat(".%s = %s(", field, swap);
+        visit(hdrExpr);
+        builder->appendFormat(".%s", field);
+        if (shift != 0)
+            builder->appendFormat(" << %d", shift);
+        builder->append(")");
+        builder->endOfStatement(true);
+    }
+    unsigned bitsInFirstByte = widthToEmit % 8;
+    if (bitsInFirstByte == 0) bitsInFirstByte = 8;
+    unsigned bitsInCurrentByte = bitsInFirstByte;
+    unsigned left = widthToEmit;
+    for (unsigned i = 0; i < (widthToEmit + 7) / 8; i++) {
+        builder->emitIndent();
+        builder->appendFormat("%s = ((char*)(&", program->byteVar.c_str());
+        visit(hdrExpr);
+        builder->appendFormat(".%s))[%d]", field, i);
+        builder->endOfStatement(true);
+        unsigned freeBits = alignment != 0 ? (8 - alignment) : 8;
+        bitsInCurrentByte = left >= 8 ? 8 : left;
+        unsigned bitsToWrite =
+                bitsInCurrentByte > freeBits ? freeBits : bitsInCurrentByte;
+        BUG_CHECK((bitsToWrite > 0) && (bitsToWrite <= 8),
+                  "invalid bitsToWrite %d", bitsToWrite);
+        builder->emitIndent();
+        if (alignment == 0 && bitsToWrite == 8) {  // write whole byte
```

Comment:
looking forward for tests that will validate this.
I got this kind of code several times wrong, so I don't trust myself to validate it by reading it.

Reply:
yeah, this part is always tricky. We'll have tests for that.